### PR TITLE
Adding unit tests for RS methods

### PR DESCRIPTION
**Pull request by rstoneback**
# Description

Addresses #6, partially

Creates unit tests for the instrument_view_through_model function

## Type of change

Please delete options that are not relevant.

- Unit Tests

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide
instructions so we can reproduce. Please also list any relevant details for
your test configuration

- Added unit tests suite based upon tests in `TestUtilsExtractModObs`

**Test Configuration**:
* macOS 10.15.4
* Python 3.9.5
* Using `listify_bug` branch on pysat

# Checklist:

- [x] Make sure you are merging into the ``develop`` (not ``main``) branch
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [ ] Add a note to ``CHANGELOG.md``, summarizing the changes

### CODE REVIEW

**Comment by rstoneback** [comment]
```
::TestUtilsExtractInstModView::test_standard_call PASSED
::TestUtilsExtractInstModView::test_bad_arg_input[2-bad_input0-Must provide inst_name as a list] PASSED
::TestUtilsExtractInstModView::test_bad_arg_input[3-bad_input1-Must provide mod_name as a list] PASSED
::TestUtilsExtractInstModView::test_bad_arg_input[2-bad_input2-Unknown instrument location index] PASSED
::TestUtilsExtractInstModView::test_bad_arg_input[3-bad_input3-Must provide the same number] PASSED
::TestUtilsExtractInstModView::test_bad_arg_input[6-bad_input4-Must provide units for each ] PASSED
::TestUtilsExtractInstModView::test_bad_arg_input[4-naname-unknown model name for datetime] PASSED
::TestUtilsExtractInstModView::test_bad_arg_input[5-naname-Unknown model time coordinate] PASSED
::TestUtilsExtractInstModView::test_bad_kwarg_input[sel_name-bad_val0-unknown_variable is not a valid model variabl] PASSED
::TestUtilsExtractInstModView::test_bad_kwarg_input[methods-bad_val1-Methods only supports "linear" or "nearest".] PASSED
::TestUtilsExtractInstModView::test_bad_kwarg_input[methods-bad_val2-Must provide interpolation selection] PASSED
::TestUtilsExtractInstModView::test_bad_kwarg_input[model_label-1-expected str instance] PASSED
::TestUtilsExtractInstModView::test_failure_for_already_ran_data PASSED
::TestUtilsExtractInstModView::test_success_for_some_already_ran_data PASSED
```

**Comment by aburrell** [comment]
So it looks like you still have some flake8 issue to address?

**Comment by rstoneback** [comment]
Thanks for catching that.

**Review by aburrell — COMMENTED** [review]
Looks like the new tests are failing.  Let me know if you're stuck or when you have them fixed 🎣

**Comment by rstoneback** [comment]
That part is a known issue. The tests require a branch over in pysat as listed in the testing setup area of the intro text :)

Using that branch sorts out the unit test issues as well as the other test failures related to minimum versions and the like. 

Seems like everything is updating and each is causing ripples in our development pond.

File: pysatModels/tests/test_utils_match.py [hunk 1]
```diff
@@ -24,7 +24,7 @@ def setup(self):
         self.filename = "%Y-%m-%d.nofile"
         self.model_kwargs = {'platform': str('pysat'),
                              'name': str('testing_xarray'),
-                             'inst_id': '12',
+                             'num_samples': '12',
```
Comment (jklenzing) [inline pysatModels/tests/test_utils_match.py]:
```suggestion
                             'num_samples': 12,
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -4,6 +4,6 @@ All notable changes to this project will be documented in this file.
 This project adheres to [Semantic Versioning](https://semver.org/).
 
 
-[0.1.0] - 2021-xx-xx
+[0.1.0] - 2021-xx-xx
```
Comment (aburrell) [inline CHANGELOG.md:7]: Remove trailing whitespace

File: pysatModels/tests/test_utils_extract.py [hunk 3]
```diff
@@ -186,3 +186,137 @@ def test_success_for_some_already_ran_data(self):
                         == self.inst.data[tcol].shape)
                 assert len(self.inst.data[tcol][
                     ~np.isnan(self.inst.data[tcol])]) > 0
+
+
+class TestUtilsExtractInstModView:
+    """ Unit tests for utils.extract.instrument_view_through_model """
```
Comment (aburrell) [inline pysatModels/tests/test_utils_extract.py:192]:
```suggestion
    """Unit tests for `utils.extract.instrument_view_through_model`."""
```

File: pysatModels/tests/test_utils_extract.py [hunk 4]
```diff
@@ -186,3 +186,137 @@ def test_success_for_some_already_ran_data(self):
...
+        ps_mod.logger.addHandler(logging.StreamHandler(self.log_capture))
+        ps_mod.logger.setLevel(logging.INFO)
+
+    def teardown(self):
+        """Runs after every method to clean up previous testing."""
+        del self.inst, self.model, self.input_args, self.out, self.model_label
+        del self.input_kwargs, self.log_capture
+
+    def test_standard_call(self):
+        """Test for successful interpolation"""
```
Comment (aburrell) [inline pysatModels/tests/test_utils_extract.py:225]:
```suggestion
        """Test for successful interpolation."""
```

File: pysatModels/tests/test_utils_extract.py [hunk 5]
```diff
@@ -186,3 +186,137 @@ def test_success_for_some_already_ran_data(self):
...
+                              (3, [], 'Must provide mod_name as a list'),
+                              (2, ['glon', 'latitude', 'altitude'],
+                               "Unknown instrument location index"),
+                              (3, ['hi'], "Must provide the same number"),
+                              (6, [], "Must provide units for each "),
+                              (4, "naname", "unknown model name for datetime"),
+                              (5, "naname", "Unknown model time coordinate")])
+    def test_bad_arg_input(self, bad_index, bad_input, err_msg):
+        """ Test for expected failure with bad input arguments
+        """
```
Comment (aburrell) [inline pysatModels/tests/test_utils_extract.py:246]:
```suggestion
        """Test for expected failure with bad input arguments."""
```

File: pysatModels/tests/test_utils_extract.py [hunk 6]
```diff
@@ -186,3 +186,137 @@ def test_success_for_some_already_ran_data(self):
...
+    @pytest.mark.parametrize("bad_key,bad_val,err_msg",
+                             [("sel_name", ["unknown_variable"],
+                               "unknown_variable is not a valid model variabl"),
+                              ("methods", ["not_a_method"],
+                               'Methods only supports "linear" or "nearest".'),
+                              ("methods", ['linear', 'linear', 'linear'],
+                               "Must provide interpolation selection"),
+                              ("model_label", 1, "expected str instance")])
+    def test_bad_kwarg_input(self, bad_key, bad_val, err_msg):
+        """ Test for expected failure with bad kwarg input """
```
Comment (aburrell) [inline pysatModels/tests/test_utils_extract.py:264]:
```suggestion
        """Test for expected failure with bad kwarg input."""
```

File: pysatModels/tests/test_utils_extract.py [hunk 7]
```diff
@@ -186,3 +186,137 @@ def test_success_for_some_already_ran_data(self):
...
+        self.input_kwargs[bad_key] = bad_val
+
+        with pytest.raises((ValueError, TypeError)) as err:
+            extract.instrument_view_through_model(*self.input_args,
+                                                  **self.input_kwargs)
+
+        assert str(err.value.args[0]).find(err_msg) >= 0
+
+    def test_failure_for_already_ran_data(self):
+        """ Test the failure for all model variables already extracted """
```
Comment (aburrell) [inline pysatModels/tests/test_utils_extract.py:274]:
```suggestion
        """Test the failure for all model variables already extracted."""
```

File: pysatModels/tests/test_utils_extract.py [hunk 8]
```diff
@@ -186,3 +186,137 @@ def test_success_for_some_already_ran_data(self):
...
+                                                  **self.input_kwargs)
+
+            self.out = self.log_capture.getvalue()
+            assert self.out.find('model data already interpolated') >= 0
+
+        assert str(err.value.args[0]).find(
+            'instrument object already contains all model data') >= 0
+
+    def test_success_for_some_already_ran_data(self):
+        """ Test the success for some model variables already extracted """
```
Comment (aburrell) [inline pysatModels/tests/test_utils_extract.py:294]:
```suggestion
        """Test the success for some model variables already extracted."""
```

File: pysatModels/tests/test_utils_extract.py [hunk 9]
```diff
@@ -186,3 +186,137 @@ def test_success_for_some_already_ran_data(self):
...
+        all_sel.append('dummy72')
+        self.input_kwargs['model_label'] = self.model_label
+
+        self.model['dummy72'] = self.model['dummy2']
+
+        # Run through twice
+        for i, selected in enumerate([all_sel[1:], all_sel]):
+            self.input_kwargs['sel_name'] = selected
+            self.input_kwargs['methods'] = ['linear'] * len(selected)
+            self.out = \
```
Comment (aburrell) [inline pysatModels/tests/test_utils_extract.py:306]: Break on open paren.

File: pysatModels/utils/extract.py [hunk 10]
```diff
@@ -250,21 +250,27 @@ def instrument_view_through_model(inst, model, inst_name, mod_name,
...
 
     # Ensure the coordinate and data variable names are array-like
-    inst_name = np.asarray(inst_name)
-    mod_name = np.asarray(mod_name)
-    method = np.asarray(methods)
+    inst_name = pyutils.listify(inst_name)
+    mod_name = pyutils.listify(mod_name)
+    methods = pyutils.listify(methods)
 
     # interp over all vars if None provided
```
Comment (aburrell) [inline pysatModels/utils/extract.py:257]:
```suggestion
    # Interpolate over all variables if None provided
```

**Review by aburrell — COMMENTED** [review]
Also need to add "return" to unit test methods for consistent style.

**Review by aburrell — COMMENTED** [review]
Test docstrings need a scrub.  We're supposed to use imperative tense (e.g., "Create clean test environment." instead of "Creates the testing environment."), have no space before the first letter, and have a period at the end.

**Comment by rstoneback** [comment]
Thanks @aburrell for the additional notes.

**Review by aburrell — APPROVED** [review]

**Pull request merged**